src/codegen/ast.lisp

```diff
@@ -78,48 +81,64 @@
 (defstruct (node-literal
             (:include node)
             (:constructor node-literal (type value)))
+  "Literal values like 1 or \"hello\""
   (value (required 'value) :type literal-value :read-only t))
 
 (defstruct (node-variable
             (:include node)
             (:constructor node-variable (type value)))
+  "Variables like x or y"
   (value (required 'value) :type symbol :read-only t))
 
 (defstruct (node-application
             (:include node)
             (:constructor node-application (type rator rands &key (pure nil))))
+  "Function application (f x)"
   (rator (required 'rator) :type node      :read-only t)
   (rands (required 'rands) :type node-list :read-only t)
   (pure  (required 'pure)  :type boolean   :read-only t))
 
 (defstruct (node-direct-application
             (:include node)
             (:constructor node-direct-application (type rator-type rator rands)))
+  "Fully saturated function application of a known function"
   (rator-type (required 'rator-type) :type tc:ty     :read-only t)
   (rator      (required 'rator)      :type symbol    :read-only t)
   (rands      (required 'rands)      :type node-list :read-only t))
 
 (defstruct (node-abstraction
             (:include node)
             (:constructor node-abstraction (type vars subexpr)))
-  (vars    (required 'vars)    :type symbol-list :read-only t)
+  "Lamba literals (fn (x) x)"
+  (vars    (required 'vars)    :type symbol-list        :read-only t)
   (subexpr (required 'subexpr) :type node               :read-only t))
 
+(defstruct (node-bare-abstraction
+            (:include node)
+            (:constructor node-bare-abstraction (type vars subexpr)))
+  "Lambda literals which do not need be wrapped in function-entries. This
```

Comment:
can you add a "because ..."? Just a little more insight as to why they don't need to be wrapped.